Formatting fails with pylsp

The following error is displayed:

```
Mon Jan 22 17:17:18 2024 [INFO] [toolbox-pylsp] Initialized at /var/home/kramo/.config/lite-xl/plugins/lsp/init.lua:293
Mon Jan 22 17:17:27 2024 [ERROR] /var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:87: invalid table: mixed or invalid key types at /var/home/kramo/.local/share/lite-xl/core/init.lua:1227

stack traceback:
[C]: in function 'error'
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:87: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:163: in function 'plugins.lsp.json.encode'
/var/home/kramo/.config/lite-xl/plugins/lsp/server.lua:648: in function 'plugins.lsp.server.process_requests'
/var/home/kramo/.config/lite-xl/plugins/lsp/init.lua:2073: in function </var/home/kramo/.config/lite-xl/plugins/lsp/init.lua:2048>
[C]: in function 'xpcall'
/var/home/kramo/.local/share/lite-xl/core/init.lua:1226: in function 'core.try'
(...tail calls...)
```

```
Sat Jan 27 18:14:02 2024 [INFO] [LSP/pylsp]: Adding request textDocument/formatting
 at /var/home/kramo/.config/lite-xl/plugins/lsp/init.lua:761
Sat Jan 27 18:14:02 2024 [ERROR] /var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:87: invalid table: mixed or invalid key types at /var/home/kramo/.local/share/lite-xl/core/init.lua:1227

stack traceback:
[C]: in function 'error'
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:87: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:111: in function </var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:73>
(...tail calls...)
/var/home/kramo/.config/lite-xl/plugins/lsp/json.lua:163: in function 'plugins.lsp.json.encode'
/var/home/kramo/.config/lite-xl/plugins/lsp/server.lua:648: in function 'plugins.lsp.server.process_requests'
/var/home/kramo/.config/lite-xl/plugins/lsp/init.lua:2073: in function </var/home/kramo/.config/lite-xl/plugins/lsp/init.lua:2048>
[C]: in function 'xpcall'
/var/home/kramo/.local/share/lite-xl/core/init.lua:1226: in function 'core.try'
(...tail calls...)
```

--- a/init.lua
+++ b/init.lua
@@ -1915,17 +1915,32 @@ function lsp.request_document_format(doc)
     servers_found = true
     local server = lsp.servers_running[name]
     if server.capabilities.documentFormattingProvider then
+      local trim_trailing_whitespace = false
+      local trim_newlines = false
+      if type(config.plugins.trimwhitespace) == "table"
+         and config.plugins.trimwhitespace.enabled
+      then
+        trim_trailing_whitespace = true
+        trim_newlines = config.plugins.trimwhitespace.trim_empty_end_lines
+      elseif config.plugins.trimwhitespace then -- Plugin enabled with true
+        trim_trailing_whitespace = true
+        trim_newlines = true
+      end
+      local indent_type, indent_size, indent_confirmed = doc:get_indent_info()
+      if not indent_confirmed then
+        indent_type, indent_size = config.tab_type, config.indent_size
+      end
       server:push_request('textDocument/formatting', {
         params = {
           textDocument = {
             uri = util.touri(core.project_absolute_path(doc.filename)),
           },
           options = {
-            tabSize = config.indent_size,
-            insertSpaces = config.tab_type == "soft",
-            trimTrailingWhitespace = config.plugins.trimwhitespace or true,
+            tabSize = indent_size,
+            insertSpaces = indent_type == "soft",
+            trimTrailingWhitespace = trim_trailing_whitespace,
             insertFinalNewline = false,
-            trimFinalNewlines = true
+            trimFinalNewlines = trim_newlines
           }
         },
         callback = function(server, response)

--- a/init.lua
+++ b/init.lua
@@ -386,8 +386,9 @@ end
 ---@param doc core.doc
 ---@param text_edit table
 ---@param is_snippet boolean
+---@param update_cursor_position boolean
 ---@return boolean True on success
-local function apply_edit(server, doc, text_edit, is_snippet)
+local function apply_edit(server, doc, text_edit, is_snippet, update_cursor_position)
   local range = nil
 
   if text_edit.range then
@@ -433,7 +434,9 @@ local function apply_edit(server, doc, text_edit, is_snippet)
   end
 
   doc:insert(line1, col1, text)
-  doc:set_selection(line2, col1+#text, line2, col1+#text)
+  if update_cursor_position then
+    doc:set_selection(line2, col1+#text, line2, col1+#text)
+  end
 
   return true
 end
@@ -513,7 +516,7 @@ local function autocomplete_onselect(index, item)
     if dv then
       local is_snippet = completion.insertTextFormat
         and completion.insertTextFormat == Server.insert_text_format.Snippet
-      local edit_applied = apply_edit(item.data.server, dv.doc, completion.textEdit, is_snippet)
+      local edit_applied = apply_edit(item.data.server, dv.doc, completion.textEdit, is_snippet, true)
       if edit_applied then
         -- Retrigger code completion if last char is a trigger
         -- this is useful for example with clangd when autocompleting
@@ -1947,8 +1950,13 @@ function lsp.request_document_format(doc)
           if response.error and response.error.message then
             log(server, "Error formatting: " .. response.error.message)
           elseif response.result and #response.result > 0 then
-            for _, result in pairs(response.result) do
-              apply_edit(server, doc, result)
+            -- Apply edits in reverse, as the ranges don't consider
+            -- the intermediate states.
+            -- Consider the TextEdits as already sorted.
+            -- If there are servers that don't sort their TextEdits,
+            -- we'll add sorting code.
+            for i=#response.result,1,-1 do
+              apply_edit(server, doc, response.result[i], false, false)
             end
             log(server, "Formatted document")
           else

